Mathlib/AlgebraicGeometry/ProjectiveSpectrum/StructureSheaf.lean

```diff
@@ -0,0 +1,419 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+
+! This file was ported from Lean 3 source module algebraic_geometry.projective_spectrum.structure_sheaf
+! leanprover-community/mathlib commit 486cb2f3bda4a67557c6285f5bd0c3348c1eea81
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.AlgebraicGeometry.ProjectiveSpectrum.Topology
+import Mathlib.Topology.Sheaves.LocalPredicate
+import Mathlib.RingTheory.GradedAlgebra.HomogeneousLocalization
+import Mathlib.AlgebraicGeometry.LocallyRingedSpace
+
+/-!
+# The structure sheaf on `projective_spectrum 𝒜`.
+
+In `Mathlib.AlgebraicGeometry.Topology`, we have given a topology on `ProjectiveSpectrum 𝒜`; in
+this file we will construct a sheaf on `ProjectiveSpectrum 𝒜`.
+
+## Notation
+- `R` is a commutative semiring;
+- `A` is a commutative ring and an `R`-algebra;
+- `𝒜 : ℕ → Submodule R A` is the grading of `A`;
+- `U` is opposite object of some open subset of `ProjectiveSpectrum.top`.
+
+## Main definitions and results
+We define the structure sheaf as the subsheaf of all dependent function
+`f : Π x : U, HomogeneousLocalization 𝒜 x` such that `f` is locally expressible as ratio of two
+elements of the *same grading*, i.e. `∀ y ∈ U, ∃ (V ⊆ U) (i : ℕ) (a b ∈ 𝒜 i), ∀ z ∈ V, f z = a / b`.
+
+* `AlgebraicGeometry.ProjectiveSpectrum.StructureSheaf.isLocallyFraction`: the predicate that
+  a dependent function is locally expressible as a ratio of two elements of the same grading.
+* `AlgebraicGeometry.ProjectiveSpectrum.StructureSheaf.sectionsSubring`: the dependent functions
+  satisfying the above local property forms a subring of all dependent functions
+  `Π x : U, HomogeneousLocalization 𝒜 x`.
+* `AlgebraicGeometry.Proj.StructureSheaf`: the sheaf with `U ↦ sectionsSubring U` and natural
+  restriction map.
+
+Then we establish that `Proj 𝒜` is a `LocallyRingedSpace`:
+* `AlgebraicGeometry.Proj.stalkIso'`: for any `x : ProjectiveSpectrum 𝒜`, the stalk of
+  `Proj.StructureSheaf` at `x` is isomorphic to `HomogeneousLocalization 𝒜 x`.
+* `AlgebraicGeometry.Proj.toLocallyRingedSpace`: `Proj` as a locally ringed space.
+
+## References
+
+* [Robin Hartshorne, *Algebraic Geometry*][Har77]
+
+
+-/
+
+
+noncomputable section
+
+namespace AlgebraicGeometry
+
+open scoped DirectSum BigOperators Pointwise
+
+open DirectSum SetLike Localization TopCat TopologicalSpace CategoryTheory Opposite
+
+variable {R A : Type _}
+
+variable [CommRing R] [CommRing A] [Algebra R A]
+
+variable (𝒜 : ℕ → Submodule R A) [GradedAlgebra 𝒜]
+
+-- porting note: this should be "local notation", but as "local notation", `at x` does not work
+-- this should be fixed before merging...
+set_option quotPrecheck false in
+notation "at " x => HomogeneousLocalization.AtPrime 𝒜
```

Comment:
This should be a local notation, but if I put `local notation`, the notation just does not work. How can we specify the correct scope of this notation?

Reply:
Made it into `local macro` and it seems to work as intended.

Reply:
#5758 is a PR to get this to pretty print

The `notation` command can't generate pretty printers if there's dot notation or `variable` variables, but `notation3` is able to at least deal with the latter.